api/v1alpha1/amaltheasession_types.go

```diff
@@ -0,0 +1,265 @@
+	// +kubebuilder:default:=8000
+	// +kubebuilder:validation:ExclusiveMinimum:=true
+	// +kubebuilder:validation:Minimum:=0
+	// The TCP port where whatever is running in the session container will listen on for connections
+	Port    int     `json:"port"`
+	Storage Storage `json:"storage,omitempty"`
+	// The abolute path for the working directory of the session container, if omitted it will use the image
+	// working directory.
+	WorkingDir string `json:"workingDir,omitempty"`
+	// +kubebuilder:default:=1000
+	// +kubebuilder:validation:Minimum:=0
+	RunAsUser int `json:"runAsUser"`
+	// +kubebuilder:default:=1000
+	// +kubebuilder:validation:Minimum:=0
+	RunAsGroup int `json:"runAsGroup"`
+	// The path where the session can be accessed. If an ingress is specified, this value must
+	// be a subpath of or identical to the ingress `pathPrefix` field.
+	URLPath string `json:"urlPath,omitempty"`
+}
+
+type Ingress struct {
+	Annotations      map[string]string `json:"annotations,omitempty"`
+	IngressClassName string            `json:"ingressClassName,omitempty"`
+	Host             string            `json:"host,omitempty"`
+	PathPrefix       string            `json:"pathPrefix,omitempty"`
+	// The name of the TLS secret, same as what is specified in a regular Kubernetes Ingress.
+	TLSSecretName string `json:"tlsSecretName,omitempty"`
+}
+
+type Storage struct {
+	ClassName string            `json:"storageClassName,omitempty"`
+	Size      resource.Quantity `json:"storageSize,omitempty"`
+	// The absolute mount path for the session volume
+	// +kubebuilder:default:=/workspace
+	MountPath string `json:"mountPath,omitempty"`
+}
+
+type CodeRepository struct {
+	// +kubebuilder:example:=repositories/project1
+	// +kubebuilder:default:="."
+	// Path relative to the session working directory where the repository should be cloned into.
+	ClonePath string `json:"clonePath,omitempty"`
+	// +kubebuilder:example:="https://github.com/SwissDataScienceCenter/renku"
+	// The HTTP url to the git repository
+	Remote string `json:"remote"`
+	// +kubebuilder:example:=main
+	// The tag, branch or commit SHA to checkout, if ommitted then will be the tip of the default branch of the repo
+	Revision string `json:"revision,omitempty"`
+	// The Kubernetes secret that contains the git configuration to be used during cloning.
+	// This can be used to inject credentials in addition to any other repo-specific Git configuration.
+	// NOTE: you have to specify the whole git config in a single key in the secret.
+	CloningGitConfigSecretRef SecretRef `json:"cloningGitConfigSecretRef,omitempty"`
+	// The Kubernetes secret that contains the git configuration to be used when the session is running.
+	// This can be used to inject credentials in addition to any other repo-specific Git configuration.
+	// NOTE: you have to specify the whole git config in a single key in the secret.
+	GitConfigSecretRef SecretRef `json:"gitConfigSecretRef,omitempty"`
```

Comment:
In the original CRD I used a dedicated Git object as I was thinking that there might be other tools that could be requested. It is unlikely that at this time code is not available through git but I do wonder if we should keep that separated so it's a field that can be swapped for something else like for example `curl` or [fossil](https://www.fossil-scm.org/home/doc/trunk/www/fossil-v-git.wiki) (note I haven't used it yet just jumped from memory).

Reply:
This is a very good point. And to follow the same pattern we have elsewhere I added a type field for the code repositories. Currently the only supported value is `git` but we can expand it in the future if we need it. And it is more uniform with the data sources and authentication sections.